fix: generalize and animate AnimatedMasonryListFooterComponent

gkartalis opened the pull request:
This PR resolves [PHIRE-1124] <!-- eg [PROJECT-XXXX] -->

### Description

During the mobile QA we noticed that when doing infinite scroll in any `Tabs.Masonry` surface the spinner was never displayed for the `first` refetch. 

This PR fixes that. Note that in order to do it we had to introduce a boolean state variable on the artworkGrids that use the legacy relay containers.

In surfaces where we use the hooks (i.e. Fairs), it is **not needed** since the `isLoadingNext` is working as expected.

Unified also the page size for all surfaces and the spinner and **bonus** fix was that with the new `AnimatedMasonryListFooterComponent` the content is no longer jumpy after the next page is rendered ⭐ .

Tested on all surfaces on both Android and iOS - Attaching before and after videos for `CollectionArtworks` and `ArtistArtworks`

|Before|After|
|---|---|
|<video src="https://github.com/user-attachments/assets/68a780e3-0285-4cf4-9323-40e9a7b2d006" width="400" />|<video src="https://github.com/user-attachments/assets/c2abc4c7-78b4-48eb-b206-06d6426390ca" width="400" />|
|<video src="https://github.com/user-attachments/assets/ecc09c86-cc6e-45c1-a124-92cbfa9fa44c" width="400" />|<video src="https://github.com/user-attachments/assets/ff9b229d-f8be-4a88-b0be-cbfc34909961" width="400" />|












<!-- Info, implementation, how to get there, before & after screenshots & videos, follow-up work, etc -->

### PR Checklist

- [x] I have tested my changes on **iOS** and **Android**.
- [x] I hid my changes behind a **[feature flag]**, or they don't need one.
- [x] I have included **screenshots** or **videos**, or I have not changed the UI.
- [x] I have added **tests**, or my changes don't require any.
- [x] I added an **[app state migration]**, or my changes do not require one.
- [x] I have documented any **follow-up work** that this PR will require, or it does not require any.
- [x] I have added a **changelog entry** below, or my changes do not require one.

### To the reviewers 👀

- [ ] I would like **at least one** of the reviewers to **run** this PR on the simulator or device.

<details><summary>Changelog updates</summary>

### Changelog updates

<!-- 📝 Please fill out at least one of these sections. -->
<!-- ⓘ 'User-facing' changes will be published as release notes. -->
<!-- ⌫ Feel free to remove sections that don't apply. -->
<!-- • Write a markdown list or just a single paragraph, but stick to plain text. -->
<!-- 📖 eg. `Enable lotsByFollowedArtists - john` or `Fix phone input misalignment - mary`. -->
<!-- 🤷‍♂️ Replace this entire block with the hashtag `#nochangelog` to avoid updating the changelog. -->
<!-- ⚠️ Prefix with `[NEEDS EXTERNAL QA]` if a change requires external QA -->

#### Cross-platform user-facing changes

- fix spinner not being displayed for first refetch on Tabbed Masonry - gkartalis

#### iOS user-facing changes

-

#### Android user-facing changes

-

#### Dev changes

-

<!-- end_changelog_updates -->

</details>

Need help with something? Have a look at our [docs], or get in touch with us.

[app state migration]: ../blob/main/docs/adding_state_migrations.md
[feature flag]: ../blob/main/docs/developing_a_feature.md
[docs]: ../blob/main/docs/README.md


[PHIRE-1124]: https://artsyproduct.atlassian.net/browse/PHIRE-1124?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

ArtsyOpenSource commented:
<!--
  0 failure: 
  0 warning: 
  
  1 markdown notices
  DangerID: danger-id-Danger;
-->



### This PR contains the following changes:

- Cross-platform user-facing changes (fix spinner not being displayed for first refetch on Tabbed Masonry - gkartalis)
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 11e06edb2c007e75234d3c8166a3d851dfb94e9a
</p>

araujobarret commented on src/app/Components/Artist/ArtistArtworks/ArtistArtworks.tsx:
```diff
@@ -31,7 +30,9 @@ import { Props as InfiniteScrollGridProps } from "app/Components/ArtworkGrids/In
 import { extractNodes } from "app/utils/extractNodes"
 import { useFeatureFlag } from "app/utils/hooks/useFeatureFlag"
 import {
+  AnimatedMasonryListFooterComponent,
```
looking at how big this file is becoming we could consider breaking it down, this new component seems to be a good candidate to be in another area

gkartalis replied:
I would say that in this PR I made it a tiny bit smaller - I wouldn't touch it more since it belongs to another team and it would be out of context for this PR

araujobarret replied:
Sorry, I meant the `app/utils/masonryHelpers`, I think we could slice some stuff from there, I'd def leave ArtistsArtworks alone 😂

gkartalis replied:
What do you mean exactly by `another area`? Like another file? Would this be for the components? ConstantS?

araujobarret replied:
Everything there is under `index.tsx`, I think we could leave the **index** as it is for now and just add the new element under:
```
|--app/utils/masonryHelpers
|---------------------------- index.tsx
|---------------------------- AnimatedMasonryListFooterComponent.tsx
```

gkartalis replied:
done!

araujobarret commented on src/app/Components/Artist/ArtistArtworks/ArtistArtworks.tsx:
```diff
@@ -133,11 +135,15 @@ const ArtworksGrid: React.FC<ArtworksGridProps> = ({
 
-  const shouldDisplaySpinner = !!artworks.length && !!relay.isLoading() && !!relay.hasMore()
+  const shouldDisplaySpinner =
+    !!isLoading && !!artworks.length && !!relay.isLoading() && !!relay.hasMore()
```
`artworks.length` is not needed here, if the result is `0` the `hasMore()` check will also return false, isn't it?

gkartalis replied:
if `artworks.length` is 0 it will exit early - same if the `isLoading` is false won't check the rest. Do you suggest a different approach?

araujobarret replied:
Good point about exiting earlier, my comment was oriented on reducing the number of checks to make it easier to read(3 checks instead of 4), perf-wise don't think it's a huge deal, but DX it could improve, will leave it to you 👍

araujobarret commented on src/app/Components/Artist/ArtistArtworks/ArtistArtworks.tsx:
```diff
@@ -133,11 +135,15 @@ const ArtworksGrid: React.FC<ArtworksGridProps> = ({
   const loadMore = useCallback(() => {
     if (relay.hasMore() && !relay.isLoading()) {
-      relay.loadMore(10)
+      setIsLoading(true)
```
maybe tag this with some comment because it shouldn't be the standard way of checking if is loading on paginated data

gkartalis replied:
comment sounds good

araujobarret reviewed (commented):
If it's for a hotfix happy to approve and move on, but wonder if we could spend a bit more time on it if we don't need to ship in this release 😅 

My main concern is that we need this extra `isLoading` state when other apps where we use pagination hooks/containers we don't

gkartalis commented:
> If it's for a hotfix happy to approve and move on, but wonder if we could spend a bit more time on it if we don't need to ship in this release 😅

It is not a hotfix. We can def spend more time on it but what do you have in mind exactly?

> My main concern is that we need this extra isLoading state when other apps where we use pagination hooks/containers we don't

This behavior has to do with how different relay legacy `isLoading()` and `hasMore()` versus the new from hooks `isLoadingNext` and `hasNext`. 

I am happy to add comments but what else do you suggest as an improvement or change?

araujobarret commented:
@gkartalis I see I see, checked the FairArtworks which has really hooks and we don't need it, so it's Relay's old remains causing the issue rather than our component setup asking for extra state control.

Fair enough from my side, just in the comment there add the reference to try to remove it as soon as these components are migrated to relay hooks, what do you think?

gkartalis commented:
@araujobarret Sounds good, will add the comment on all cases 🌟

gkartalis commented:
Done! Added comments and moved the component to its own file under masonryHelpers - I did remove the `Component` part from the name as well since it felt too much

araujobarret commented:
It looks solid, thanks for the changes!
